### PR TITLE
Csp error strings missing error code parameter 

### PR DESCRIPTION
**Pull request by ViktorHofer**
fixes https://github.com/dotnet/corefx/issues/11646

### CODE REVIEW

**Comment by bartonjs** [comment]
Wow, that was fast.  And pretty thorough-looking.

A lot of the places that you changed to hr.ToCryptographicException would have had slightly different behaviors on .NET Framework, because it was all one big method returning the HRESULT and it got an "if !success throw generic message" treatment (a la  https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/icspasymmetricalgorithm.cs,106).  I don't think it's worth rewriting all of this to go back to an HRESULT style, so your changes are good.  (They're both CryptographicException, and corefx gives more information without being an information disclosure issue, so seems fine)

https://github.com/dotnet/corefx/pull/18161/files#diff-91c6f6810a34222999754d7d7278ea34L297, on the other hand, we're throwing a HRESULT-free value (using a custom message) where I think netfx threw an exception using the OS string (https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/utils.cs,214)    (netfx `new CryptographicException(hr)` => corefx `hr.ToCryptographicException()`, since we had to subclass CryptographicException for message loading across platforms)

**Comment by ViktorHofer** [comment]
@bartonjs: PTAL

**Comment by ViktorHofer** [comment]
"Argument_InvalidValue" resource string never had any format identifier like {0}, neither in corefx nor in netfx:

> Argument_InvalidKey = Key was invalid.

but it was used here with format parameters:

CapiHelper line 802:
`throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted Data is null"));`

and here CapiHelper line 806:
`throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted data length is less than 0"));`

So I guess these parameters are also being ignored in netfx.

File: src/System.Security.Cryptography.Csp/src/Resources/Strings.resx [hunk 1]
```diff
@@ -190,4 +166,4 @@
   <data name="OpenCSP_Failed" xml:space="preserve">
     <value>OpenCSP failed with error code {0}.</value>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/Resources/Strings.resx]: I think this is now unused and can be deleted.

**Comment by bartonjs** [comment]
"Argument_InvalidValue" resource string never had any format identifier like {0}, neither in corefx nor in netfx:

@stephentoub @danmosemsft I don't know how our loc process does/will work here.  Do we want to have identical values for identical keys here as in netfx; or is the value in corefx allowed to be "better"?

**Comment by danmosemsft** [comment]
@bartonjs we no longer have a goal to share sources with desktop by xcopy, and that includes resx. We've added lots of strings anyway: from a loc point of view a change and an add comes to the same work. Also, we don't currently localize .NET Core. So I think in short we can do whatever we want, and we *should* aim to improve messages.

File: src/System.Security.Cryptography.Csp/src/Resources/Strings.resx (hunk 1 above)
Comment (ViktorHofer) [inline src/System.Security.Cryptography.Csp/src/Resources/Strings.resx]: It is used in L136 and L200. But yeah I think that error message isn't helpful at all...

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs [hunk 2]
```diff
@@ -799,11 +799,11 @@ internal static void DecryptKey(SafeKeyHandle safeKeyHandle, byte[] encryptedDat
             VerifyValidHandle(safeKeyHandle);
             if (null == encryptedData)
             {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted Data is null"));
+                throw new CryptographicException(SR.Argument_InvalidValue + " Encrypted Data is null");
             }
             if (encryptedDataLength < 0)
             {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted data length is less than 0"));
+                throw new CryptographicException(SR.Argument_InvalidValue + " Encrypted data length is less than 0");
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: These changes aren't localizable.  Which sent me on a hunt.

In netfx these two things were checked inside the CLR via a PRECONDITION tests in a QCall.  That turns out to be equivalent to Debug.Assert().  So the two if/throws here should just be changed to Debug.Assert(the opposite of what the if says).

Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: The extra literal string can happily be the message to Debug.Assert, though :smile:

File: src/System.Security.Cryptography.Csp/src/Resources/Strings.resx (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/Resources/Strings.resx]: Ah, they should probably both be hr.ToCryptographicException() :smile:

**Comment by ViktorHofer** [comment]
Updated src and fixed tests.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs [hunk 3]
```diff
@@ -797,14 +797,9 @@ internal static void SetPersistKeyInCsp(SafeProvHandle safeProvHandle, bool fPer
...
-            if (null == encryptedData)
-            {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted Data is null"));
-            }
-            if (encryptedDataLength < 0)
-            {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "Encrypted data length is less than 0"));
-            }
+            Debug.Assert(encryptedData != null, SR.Argument_InvalidValue, "Encrypted Data is null");
+            Debug.Assert(encryptedDataLength >= 0, SR.Argument_InvalidValue, "Encrypted data length is less than 0");
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: Just `Debug.Assert(encryptedDat != null, "encryptedData is null")`. (and similar) The Argument_InvalidValue doesn't really play a role here.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs [hunk 4]
```diff
@@ -862,20 +857,15 @@ internal static void DecryptKey(SafeKeyHandle safeKeyHandle, byte[] encryptedDat
...
-            if (null == pbKey)
-            {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "pbKey is null"));
-            }
-            if (cbKey < 0)
-            {
-                throw new CryptographicException(SR.Format(SR.Argument_InvalidValue, "cbKey is less than 0"));
-            }
+            Debug.Assert(pbKey != null, SR.Argument_InvalidValue, "pbKey is null");
+            Debug.Assert(cbKey >= 0, SR.Argument_InvalidValue, "cbKey is less than 0");
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: Oh, yup, Debug.Assert is correct here; good call.  But the messages should be fixed.  I didn't mention it above, but the length one can be done more helpfully:

`Debug.Assert(cbKey >= 0, $"cbKey is less than 0 ({cbKey});"`

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs [hunk 5]
```diff
@@ -698,7 +698,7 @@ private static void ValidateCspFlags(CspProviderFlags flags)
                                             CspProviderFlags.UseUserProtectedKey);
                 if ((flags & keyFlags) != CspProviderFlags.NoFlags)
                 {
-                    throw new ArgumentException(SR.Format(SR.Argument_InvalidValue, Convert.ToString(flags)));
+                    throw new ArgumentException(SR.Argument_InvalidValue, nameof(flags));
```
Comment (danmosemsft) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: Convert.ToString(flags) produces something like "UseExistingKey, CreateEphemeralKey" that could potentially be helpful. is it worth including it? It's too bad ArgumentExceptoin doesn't have a field for "actual value"

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs [hunk 6]
```diff
@@ -738,7 +738,7 @@ private static void ValidateCspFlags(CspProviderFlags flags)
                 (keyType == CspAlgorithmType.Dss && dwAlgId != CALG_DSS_SIGN))
             {
                 hKey.Dispose();
-                throw new CryptographicException(SR.Format(SR.Cryptography_CSP_WrongKeySpec, Convert.ToString(keyType)));
+                throw new CryptographicException(SR.Cryptography_CSP_WrongKeySpec);
```
Comment (danmosemsft) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: Same here also, is it worth formatting `Convert.ToString(keyType)` into the message.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs (hunk 5 above)
Comment (ViktorHofer) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs]: +1 for ArgumentException needs an actual value 👍 Yeah I will use the Format + argumentname so a mix of both. thanks Dan!

**Comment by ViktorHofer** [comment]
Had to resolve a few merge conflicts as @steveharter pushed a pretty substantial commit yesterday. Should be fine now.

**Pull request merged**